File: src/client/java/com/cxxcxx/zinecraft/core/client/weapon/tacz/TaczBedrockModel.java
package com.cxxcxx.zinecraft.core.client.weapon.tacz;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.google.gson.stream.JsonReader;
import net.minecraft.resources.ResourceLocation;

import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.util.*;

record TaczVector(float x, float y, float z) {
  static final TaczVector ZERO = new TaczVector(0, 0, 0);
  static final TaczVector ONE = new TaczVector(1, 1, 1);

  TaczVector add(TaczVector other) {
    return new TaczVector(x + other.x, y + other.y, z + other.z);
  }

  TaczVector subtract(TaczVector other) {
    return new TaczVector(x - other.x, y - other.y, z - other.z);
  }

  TaczVector multiply(float scale) {
    return new TaczVector(x * scale, y * scale, z * scale);
  }
}

record TaczFaceUv(float u, float v, float width, float height) {
}

record TaczCube(TaczVector origin, TaczVector size, TaczVector pivot, TaczVector rotation, float inflate,
                Map<String, TaczFaceUv> faces) {
}

record TaczBone(String name, String parent, TaczVector pivot, TaczVector rotation, List<TaczCube> cubes) {
}

record TaczBedrockModel(float textureWidth, float textureHeight, List<TaczBone> bones) {
}

record TaczBoneTransform(TaczVector position, TaczVector rotation, TaczVector scale) {
  static final TaczBoneTransform IDENTITY = new TaczBoneTransform(TaczVector.ZERO, TaczVector.ZERO, TaczVector.ONE);
}

record TaczAnimationClip(float length, boolean loop, Map<String, TaczAnimatedBone> bones,
                         List<TaczAnimationSound> sounds) {
  Map<String, TaczBoneTransform> sample(float timeSeconds) {
    float time = loop && length > 0 ? timeSeconds % length : Math.min(timeSeconds, length);
    Map<String, TaczBoneTransform> result = new LinkedHashMap<>();
    bones.forEach((name, bone) -> result.put(name, new TaczBoneTransform(
        bone.position().sample(time, TaczVector.ZERO), bone.rotation().sample(time, TaczVector.ZERO),
        bone.scale().sample(time, TaczVector.ONE))));
    return result;
  }
}

record TaczAnimatedBone(TaczKeyframes position, TaczKeyframes rotation, TaczKeyframes scale) {
}

record TaczAnimationSound(float time, ResourceLocation id) {
}

record TaczKeyframe(float time, TaczVector value, boolean catmullRom) {
}

record TaczKeyframes(List<TaczKeyframe> values) {
  private static TaczVector catmull(TaczVector p0, TaczVector p1, TaczVector p2, TaczVector p3, float t) {
    float t2 = t * t;
    float t3 = t2 * t;
    return p1.multiply(2).add(p2.subtract(p0).multiply(t))
        .add(p0.multiply(2).subtract(p1.multiply(5)).add(p2.multiply(4)).subtract(p3).multiply(t2))
        .add(p3.subtract(p0).add(p1.subtract(p2).multiply(3)).multiply(t3)).multiply(0.5f);
  }

  TaczVector sample(float time, TaczVector fallback) {
    if (values.isEmpty()) return fallback;
    if (values.size() == 1 || time <= values.getFirst().time()) return values.getFirst().value();
    if (time >= values.getLast().time()) return values.getLast().value();
    int upper = 1;
    while (upper < values.size() && values.get(upper).time() < time) upper++;
    TaczKeyframe start = values.get(upper - 1);
    TaczKeyframe end = values.get(upper);
    float alpha = Math.clamp((time - start.time()) / (end.time() - start.time()), 0, 1);
    if (!end.catmullRom()) return start.value().multiply(1 - alpha).add(end.value().multiply(alpha));
    TaczVector before = values.get(Math.max(0, upper - 2)).value();
    TaczVector after = values.get(Math.min(values.size() - 1, upper + 1)).value();
    return catmull(before, start.value(), end.value(), after, alpha);
  }
}

final class TaczBedrockParser {
  private TaczBedrockParser() {
  }

  static TaczBedrockModel model(InputStream input) {
    JsonObject root = json(input);
    var geometries = root.getAsJsonArray("minecraft:geometry");
    if (geometries == null || geometries.isEmpty()) throw new IllegalStateException("Missing minecraft:geometry");
    JsonObject geometry = geometries.get(0).getAsJsonObject();
    JsonObject description = geometry.getAsJsonObject("description");
    List<TaczBone> bones = new ArrayList<>();
    var boneArray = geometry.getAsJsonArray("bones");
    if (boneArray != null) for (JsonElement element : boneArray) {
      JsonObject bone = element.getAsJsonObject();
      TaczVector pivot = vector(bone, "pivot", TaczVector.ZERO);
      List<TaczCube> cubes = new ArrayList<>();
      var cubeArray = bone.getAsJsonArray("cubes");
      if (cubeArray != null) for (JsonElement cube : cubeArray) cubes.add(cube(cube.getAsJsonObject(), pivot));
      bones.add(new TaczBone(bone.get("name").getAsString(), string(bone, "parent"), pivot,
          vector(bone, "rotation", TaczVector.ZERO), List.copyOf(cubes)));
    }
    return new TaczBedrockModel(decimal(description, "texture_width", 16),
        decimal(description, "texture_height", 16), List.copyOf(bones));
  }

  static Map<String, TaczAnimationClip> animations(InputStream input) {
    JsonObject animations = json(input).getAsJsonObject("animations");
    if (animations == null) return Map.of();
    Map<String, TaczAnimationClip> result = new LinkedHashMap<>();
    animations.entrySet().forEach(entry -> {
      JsonObject clip = entry.getValue().getAsJsonObject();
      JsonElement loopValue = clip.get("loop");
      // Bedrock's "hold_on_last_frame" is a one-shot clip that clamps at its end, not a loop.
      boolean loop = loopValue != null && loopValue.isJsonPrimitive()
          && loopValue.getAsJsonPrimitive().isBoolean() && loopValue.getAsBoolean();
      Map<String, TaczAnimatedBone> animatedBones = new LinkedHashMap<>();
      JsonObject bones = clip.getAsJsonObject("bones");
      if (bones != null) bones.entrySet().forEach(boneEntry -> {
        JsonObject bone = boneEntry.getValue().getAsJsonObject();
        animatedBones.put(boneEntry.getKey(), new TaczAnimatedBone(
            keyframes(bone.get("position"), TaczVector.ZERO),
            keyframes(bone.get("rotation"), TaczVector.ZERO),
            keyframes(bone.get("scale"), TaczVector.ONE)));
      });
      List<TaczAnimationSound> sounds = new ArrayList<>();
      JsonObject soundEffects = clip.getAsJsonObject("sound_effects");
      if (soundEffects != null) soundEffects.entrySet().forEach(sound -> {
        try {
          if (!sound.getValue().isJsonObject()) return;
          ResourceLocation id = ResourceLocation.tryParse(string(sound.getValue().getAsJsonObject(), "effect"));
          if (id != null) sounds.add(new TaczAnimationSound(Float.parseFloat(sound.getKey()), id));
        } catch (RuntimeException ignored) {
        }
      });
      sounds.sort(Comparator.comparing(TaczAnimationSound::time));
      result.put(entry.getKey(), new TaczAnimationClip(decimal(clip, "animation_length", 0), loop,
          Map.copyOf(animatedBones), List.copyOf(sounds)));
    });
    return Map.copyOf(result);
  }

  private static TaczCube cube(JsonObject json, TaczVector bonePivot) {
    TaczVector size = vector(json, "size", TaczVector.ZERO);
    JsonElement uv = json.get("uv");
    Map<String, TaczFaceUv> faces = new LinkedHashMap<>();
    if (uv != null && uv.isJsonObject()) uv.getAsJsonObject().entrySet().forEach(entry -> {
      if (!entry.getValue().isJsonObject()) return;
      JsonObject data = entry.getValue().getAsJsonObject();
      float[] origin = vector2(data, "uv");
      float[] extent = vector2(data, "uv_size");
      if (origin != null && extent != null)
        faces.put(entry.getKey(), new TaczFaceUv(origin[0], origin[1], extent[0], extent[1]));
    });
    else if (uv != null && uv.isJsonArray() && uv.getAsJsonArray().size() >= 2) {
      faces.putAll(boxUv(uv.getAsJsonArray().get(0).getAsFloat(), uv.getAsJsonArray().get(1).getAsFloat(), size));
    }
    return new TaczCube(vector(json, "origin", TaczVector.ZERO), size, vector(json, "pivot", bonePivot),
        vector(json, "rotation", TaczVector.ZERO), decimal(json, "inflate", 0), Map.copyOf(faces));
  }

  private static Map<String, TaczFaceUv> boxUv(float u, float v, TaczVector size) {
    return Map.of("west", new TaczFaceUv(u, v + size.z(), size.z(), size.y()),
        "north", new TaczFaceUv(u + size.z(), v + size.z(), size.x(), size.y()),
        "east", new TaczFaceUv(u + size.z() + size.x(), v + size.z(), size.z(), size.y()),
        "south", new TaczFaceUv(u + size.z() * 2 + size.x(), v + size.z(), size.x(), size.y()),
        "up", new TaczFaceUv(u + size.z(), v, size.x(), size.z()),
        "down", new TaczFaceUv(u + size.z() + size.x(), v, size.x(), size.z()));
  }

  private static TaczKeyframes keyframes(JsonElement element, TaczVector fallback) {
    if (element == null) return new TaczKeyframes(List.of());
    if (element.isJsonArray()) return new TaczKeyframes(List.of(new TaczKeyframe(0, vector(element, fallback), false)));
    if (!element.isJsonObject()) return new TaczKeyframes(List.of());
    List<TaczKeyframe> values = new ArrayList<>();
    element.getAsJsonObject().entrySet().forEach(entry -> {
      try {
        float timestamp = Float.parseFloat(entry.getKey());
        JsonElement value = entry.getValue();
        JsonElement vector = value.isJsonObject()
            ? value.getAsJsonObject().has("post") ? value.getAsJsonObject().get("post") : value.getAsJsonObject().get("pre")
            : value;
        String mode = value.isJsonObject() ? string(value.getAsJsonObject(), "lerp_mode") : null;
        values.add(new TaczKeyframe(timestamp, vector(vector, fallback), "catmullrom".equals(mode)));
      } catch (RuntimeException ignored) {
      }
    });
    values.sort(Comparator.comparing(TaczKeyframe::time));
    return new TaczKeyframes(List.copyOf(values));
  }

  private static JsonObject json(InputStream input) {
    JsonReader reader = new JsonReader(new InputStreamReader(input, StandardCharsets.UTF_8));
    reader.setLenient(true);
    return JsonParser.parseReader(reader).getAsJsonObject();
  }

  private static TaczVector vector(JsonObject json, String key, TaczVector fallback) {
    return json == null ? fallback : vector(json.get(key), fallback);
  }

  private static TaczVector vector(JsonElement element, TaczVector fallback) {
    if (element == null || !element.isJsonArray() || element.getAsJsonArray().size() < 3) return fallback;
    var array = element.getAsJsonArray();
    return new TaczVector(array.get(0).getAsFloat(), array.get(1).getAsFloat(), array.get(2).getAsFloat());
  }

  private static float[] vector2(JsonObject json, String key) {
    var array = json.getAsJsonArray(key);
    return array == null || array.size() < 2 ? null : new float[]{array.get(0).getAsFloat(), array.get(1).getAsFloat()};
  }

  private static String string(JsonObject json, String key) {
    if (json == null) return null;
    JsonElement value = json.get(key);
    return value != null && value.isJsonPrimitive() ? value.getAsString() : null;
  }

  private static float decimal(JsonObject json, String key, float fallback) {
    if (json == null) return fallback;
    JsonElement value = json.get(key);
    return value != null && value.isJsonPrimitive() ? value.getAsFloat() : fallback;
  }
}

File: src/client/kotlin/com/cxxcxx/zinecraft/core/client/weapon/tacz/TaczBedrockModel.kt
package com.cxxcxx.zinecraft.core.client.weapon.tacz

import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import com.google.gson.stream.JsonReader
import java.io.InputStream
import java.io.InputStreamReader
import java.nio.charset.StandardCharsets
import net.minecraft.resources.ResourceLocation

internal data class TaczVector(val x: Float, val y: Float, val z: Float) {
  operator fun plus(other: TaczVector) = TaczVector(x + other.x, y + other.y, z + other.z)
  operator fun minus(other: TaczVector) = TaczVector(x - other.x, y - other.y, z - other.z)
  operator fun times(scale: Float) = TaczVector(x * scale, y * scale, z * scale)

  companion object {
    val ZERO = TaczVector(0f, 0f, 0f)
    val ONE = TaczVector(1f, 1f, 1f)
  }
}

internal data class TaczFaceUv(val u: Float, val v: Float, val width: Float, val height: Float)

internal data class TaczCube(
  val origin: TaczVector,
  val size: TaczVector,
  val pivot: TaczVector,
  val rotation: TaczVector,
  val inflate: Float,
  val faces: Map<String, TaczFaceUv>
)

internal data class TaczBone(
  val name: String,
  val parent: String?,
  val pivot: TaczVector,
  val rotation: TaczVector,
  val cubes: List<TaczCube>
)

internal data class TaczBedrockModel(
  val textureWidth: Float,
  val textureHeight: Float,
  val bones: List<TaczBone>
)

internal data class TaczBoneTransform(
  val position: TaczVector = TaczVector.ZERO,
  val rotation: TaczVector = TaczVector.ZERO,
  val scale: TaczVector = TaczVector.ONE
)

internal data class TaczAnimationClip(
  val length: Float,
  val loop: Boolean,
  val bones: Map<String, TaczAnimatedBone>,
  val sounds: List<TaczAnimationSound> = emptyList()
) {
  fun sample(timeSeconds: Float): Map<String, TaczBoneTransform> {
    val time = if (loop && length > 0f) timeSeconds % length else timeSeconds.coerceAtMost(length)
    return bones.mapValues { (_, bone) ->
      TaczBoneTransform(
        bone.position.sample(time, TaczVector.ZERO),
        bone.rotation.sample(time, TaczVector.ZERO),
        bone.scale.sample(time, TaczVector.ONE)
      )
    }
  }
}

internal data class TaczAnimatedBone(
  val position: TaczKeyframes,
  val rotation: TaczKeyframes,
  val scale: TaczKeyframes
)

internal data class TaczAnimationSound(val time: Float, val id: ResourceLocation)

internal data class TaczKeyframe(val time: Float, val value: TaczVector, val catmullRom: Boolean)

internal data class TaczKeyframes(val values: List<TaczKeyframe>) {
  fun sample(time: Float, fallback: TaczVector): TaczVector {
    if (values.isEmpty()) return fallback
    if (values.size == 1 || time <= values.first().time) return values.first().value
    if (time >= values.last().time) return values.last().value
    val upper = values.indexOfFirst { it.time >= time }.coerceAtLeast(1)
    val startFrame = values[upper - 1]
    val endFrame = values[upper]
    val startTime = startFrame.time
    val endTime = endFrame.time
    val alpha = ((time - startTime) / (endTime - startTime)).coerceIn(0f, 1f)
    if (!endFrame.catmullRom) return startFrame.value * (1f - alpha) + endFrame.value * alpha
    val before = values.getOrElse(upper - 2) { startFrame }.value
    val after = values.getOrElse(upper + 1) { endFrame }.value
    return catmull(before, startFrame.value, endFrame.value, after, alpha)
  }

  private fun catmull(p0: TaczVector, p1: TaczVector, p2: TaczVector, p3: TaczVector, t: Float): TaczVector {
    val t2 = t * t
    val t3 = t2 * t
    return (p1 * 2f + (p2 - p0) * t + (p0 * 2f - p1 * 5f + p2 * 4f - p3) * t2 +
        (p3 - p0 + (p1 - p2) * 3f) * t3) * 0.5f
  }
}

internal object TaczBedrockParser {
  fun model(input: InputStream): TaczBedrockModel {
    val root = json(input)
    val geometry = root.getAsJsonArray("minecraft:geometry")?.firstOrNull()?.asJsonObject
      ?: error("Missing minecraft:geometry")
    val description = geometry.getAsJsonObject("description")
    val bones = geometry.getAsJsonArray("bones")?.map { element ->
      val bone = element.asJsonObject
      val pivot = bone.vector("pivot", TaczVector.ZERO)
      TaczBone(
        name = bone.get("name").asString,
        parent = bone.string("parent"),
        pivot = pivot,
        rotation = bone.vector("rotation", TaczVector.ZERO),
        cubes = bone.getAsJsonArray("cubes")?.map { cubeElement -> cube(cubeElement.asJsonObject, pivot) }
          ?: emptyList()
      )
    } ?: emptyList()
    return TaczBedrockModel(
      textureWidth = description?.float("texture_width", 16f) ?: 16f,
      textureHeight = description?.float("texture_height", 16f) ?: 16f,
      bones = bones
    )
  }

  fun animations(input: InputStream): Map<String, TaczAnimationClip> {
    val animations = json(input).getAsJsonObject("animations") ?: return emptyMap()
    return animations.entrySet().associate { (name, element) ->
      val clip = element.asJsonObject
      val loopValue = clip.get("loop")
      val loop =
        loopValue?.let { it.isJsonPrimitive && (it.asJsonPrimitive.isBoolean && it.asBoolean || it.asString == "hold_on_last_frame") }
          ?: false
      val animatedBones = clip.getAsJsonObject("bones")?.entrySet()?.associate { (boneName, boneValue) ->
        val bone = boneValue.asJsonObject
        boneName to TaczAnimatedBone(
          keyframes(bone.get("position"), TaczVector.ZERO),
          keyframes(bone.get("rotation"), TaczVector.ZERO),
          keyframes(bone.get("scale"), TaczVector.ONE)
        )
      } ?: emptyMap()
      val sounds = clip.getAsJsonObject("sound_effects")?.entrySet()?.mapNotNull { (time, value) ->
        val effect = value.takeIf(JsonElement::isJsonObject)?.asJsonObject?.string("effect") ?: return@mapNotNull null
        val id = ResourceLocation.tryParse(effect) ?: return@mapNotNull null
        (time.toFloatOrNull() ?: return@mapNotNull null) to id
      }?.sortedBy(Pair<Float, ResourceLocation>::first)?.map { TaczAnimationSound(it.first, it.second) } ?: emptyList()
      name to TaczAnimationClip(clip.float("animation_length", 0f), loop, animatedBones, sounds)
    }
  }

  private fun cube(json: JsonObject, bonePivot: TaczVector): TaczCube {
    val size = json.vector("size", TaczVector.ZERO)
    val uv = json.get("uv")
    return TaczCube(
      origin = json.vector("origin", TaczVector.ZERO),
      size = size,
      pivot = json.vector("pivot", bonePivot),
      rotation = json.vector("rotation", TaczVector.ZERO),
      inflate = json.float("inflate", 0f),
      faces = when {
        uv == null -> emptyMap()
        uv.isJsonObject -> uv.asJsonObject.entrySet().mapNotNull { (face, value) ->
          val data = value.takeIf(JsonElement::isJsonObject)?.asJsonObject ?: return@mapNotNull null
          val origin = data.vector2("uv") ?: return@mapNotNull null
          val extent = data.vector2("uv_size") ?: return@mapNotNull null
          face to TaczFaceUv(origin.first, origin.second, extent.first, extent.second)
        }.toMap()

        uv.isJsonArray -> boxUv(uv.asJsonArray[0].asFloat, uv.asJsonArray[1].asFloat, size)
        else -> emptyMap()
      }
    )
  }

  private fun boxUv(u: Float, v: Float, size: TaczVector): Map<String, TaczFaceUv> = mapOf(
    "west" to TaczFaceUv(u, v + size.z, size.z, size.y),
    "north" to TaczFaceUv(u + size.z, v + size.z, size.x, size.y),
    "east" to TaczFaceUv(u + size.z + size.x, v + size.z, size.z, size.y),
    "south" to TaczFaceUv(u + size.z * 2 + size.x, v + size.z, size.x, size.y),
    "up" to TaczFaceUv(u + size.z, v, size.x, size.z),
    "down" to TaczFaceUv(u + size.z + size.x, v, size.x, size.z)
  )

  private fun keyframes(element: JsonElement?, fallback: TaczVector): TaczKeyframes {
    if (element == null) return TaczKeyframes(emptyList())
    if (element.isJsonArray) return TaczKeyframes(listOf(TaczKeyframe(0f, element.vector(fallback), false)))
    if (!element.isJsonObject) return TaczKeyframes(emptyList())
    val values = element.asJsonObject.entrySet().mapNotNull { (time, value) ->
      val timestamp = time.toFloatOrNull() ?: return@mapNotNull null
      val vectorElement =
        if (value.isJsonObject) value.asJsonObject.get("post") ?: value.asJsonObject.get("pre") else value
      val mode = value.takeIf(JsonElement::isJsonObject)?.asJsonObject?.string("lerp_mode")
      TaczKeyframe(timestamp, vectorElement.vector(fallback), mode == "catmullrom")
    }.sortedBy(TaczKeyframe::time)
    return TaczKeyframes(values)
  }

  private fun json(input: InputStream): JsonObject = InputStreamReader(input, StandardCharsets.UTF_8).use { reader ->
    JsonParser.parseReader(JsonReader(reader).apply { isLenient = true }).asJsonObject
  }

  private fun JsonObject.vector(key: String, fallback: TaczVector): TaczVector = get(key)?.vector(fallback) ?: fallback
  private fun JsonElement.vector(fallback: TaczVector): TaczVector {
    if (!isJsonArray || asJsonArray.size() < 3) return fallback
    return TaczVector(asJsonArray[0].asFloat, asJsonArray[1].asFloat, asJsonArray[2].asFloat)
  }

  private fun JsonObject.vector2(key: String): Pair<Float, Float>? {
    val array = getAsJsonArray(key) ?: return null
    if (array.size() < 2) return null
    return array[0].asFloat to array[1].asFloat
  }

  private fun JsonObject.string(key: String): String? = get(key)?.takeIf { it.isJsonPrimitive }?.asString
  private fun JsonObject.float(key: String, fallback: Float): Float =
    get(key)?.takeIf { it.isJsonPrimitive }?.asFloat ?: fallback
}
